[syntax-errors] `nonlocal` declaration at module level

abhijeetbodas2001 opened the pull request:
## Summary

Part of #17412

Add a new compile-time syntax error for detecting `nonlocal` declarations at a module level.

## Test Plan

- Added new inline tests for the syntax error
- Updated existing tests for `nonlocal` statement parsing to be inside a function scope

abhijeetbodas2001 commented:
@dhruvmanila could you review this?

ntBre commented on crates/ruff_python_parser/src/parser/statement.rs:
```diff
@@ -898,30 +898,18 @@ impl<'src> Parser<'src> {
         let start = self.node_start();
         self.bump(TokenKind::Nonlocal);
 
-        // test_err nonlocal_stmt_trailing_comma
-        // nonlocal ,
-        // nonlocal x,
-        // nonlocal x, y,
```
I think we should leave these as inline tests. You can still nest them in functions like in the `test_case`s:

```rust
        // test_err nonlocal_stmt_trailing_comma
        // def _():
        //     nonlocal ,
        //     nonlocal x,
        //     nonlocal x, y,
```

ntBre replied:
Actually, I think if you implemented the test version of `in_module_scope` as `self.scopes.is_empty()` you could even add a `test_ok` inline case for the new error and exclusively use inline tests.

https://github.com/astral-sh/ruff/blob/e7f38fe74ba642cbd4cf286d8524f4172cbe30e2/crates/ruff_python_parser/tests/fixtures.rs#L540-L542

We currently only track function scopes in the test visitor, but I think that's okay. It wouldn't be hard to add support for classes, if we really wanted to test that too.

abhijeetbodas2001 replied:
Ah, didn't know there's a separate visitor for running unit-tests.
I've reset the existing test to remain inline, and enclosed them in a function scope like you suggested.

ntBre reviewed (commented):
Awesome, thanks! I had one suggestion about where to put tests, but this looks great to me.

dhruvmanila commented:
(I'll leave this to Brent)

abhijeetbodas2001 commented:
@ntBre Thanks for the review. I've addressed your comment and pushed again.

I have one question -- since it looks like this repo follows the squash+merge git workflow, should I be updating the PR description every time I push after addressing review comments (since this description will end up in the commit message)?

abhijeetbodas2001 commented:
(Resolved conflicts in the last push)

ntBre reviewed (approved):
This looks great, thanks!

ntBre commented:
> I have one question -- since it looks like this repo follows the squash+merge git workflow, should I be updating the PR description every time I push after addressing review comments (since this description will end up in the commit message)?

Great question! I would say not to worry _too_ much about it, but it is definitely nice to keep it updated if things change substantially. Your edits here look perfect.

abhijeetbodas2001 commented:
> Great question! I would say not to worry too much about it, but it is definitely nice to keep it updated if things change substantially. Your edits here look perfect.

Got it, thanks!

Also wanted to mention that your detailed issue write up was really helpful from a new contributor (ie, me) POV, so thanks for that as well :smile:

ntBre commented:
Thank you, that's great to hear! And thanks for the great PRs!
